File: src/first_pkg/src/service_server.cpp
#include "ros/ros.h"
#include "first_pkg/magnitude.h"

bool magnitudeCalculation(first_pkg::magnitude::Request &req, first_pkg::magnitude::Response &resp) {
    // Mengolah data request kedalam respon
    resp.magnitude = std::sqrt(req.komponenX * req.komponenX + req.komponenY * req.komponenY + req.komponenZ * req.komponenZ);

    // Mencetak hasil olahan data
    ROS_INFO("\nNilai komponen X: %lf", (float)req.komponenX);
    ROS_INFO("\nNilai komponen Y: %lf", (float)req.komponenY);
    ROS_INFO("\nNilai komponen Z: %lf", (float)req.komponenZ);
    ROS_INFO("\nMagnitude: %lf", (float)resp.magnitude);
    return true;
}

int main(int argc, char** argv) {
    // Inisialisasi Node
    ros::init(argc, argv, "service_server");

    // Inisialisasi Handler Node
    ros::NodeHandle nh;

    // Insialisasi Server dari Handler
    ros::ServiceServer server = nh.advertiseService("magnitude", magnitudeCalculation);
    ROS_INFO("Service Server ready to receive requests.");
    ros::spin();

    return 0;
}

File: src/first_pkg/script/service_server.py
#!/usr/bin/env python3

import rospy
from first_pkg.srv import magnitude,magnitudeResponse
import math

def magnitude_calculation(req):
    # Mengolah & mencetak data
    magnitude = math.sqrt(req.komponenX**2 + req.komponenY**2 + req.komponenZ**2)
    rospy.loginfo("Received request: Komponen X=%f, Komponen Y=%f, Komponen Z=%f" % (req.komponenX, req.komponenY, req.komponenZ))
    rospy.loginfo("Returning magnitude: %f" % magnitude)
    return magnitudeResponse(magnitude)

def service_server():
    # Inisialisasi Node
    rospy.init_node("service_server")
    
    # Insialisasi Server
    s = rospy.Service('magnitude_calculation_python',magnitude,magnitude_calculation)
    rospy.loginfo("Magnitude Calculation Server is ready.")
    rospy.spin()

if __name__ == '__main__':
    service_server()